test/test_genome.py

```diff
@@ -346,89 +347,154 @@ def test_is_gene_in_output_region(rng_seed):
+        )  # only one gene per output can be mutated
+        if n_inputs == 1:
+            n_immutable_genes += (
+                n_row * 2
+            )  # input gene addresses in the first hidden layer can't be mutated in that case
+        return n_immutable_genes
+
+    def count_mutations(prev_dna, new_dna):
+        counter = 0
+        for idx in range(len(prev_dna)):
+            if prev_dna[idx] != new_dna[idx]:
+                counter += 1
+        return counter
+
+    n_immutable_genes = count_n_immutable_genes(n_inputs, n_outputs, n_rows)
+    n_mutations_mean_expected = mutation_rate * (len(genome.dna) - n_immutable_genes)
+    n_mutations_std_expected = np.sqrt(
+        (len(genome.dna) - n_immutable_genes) * mutation_rate * (1 - mutation_rate)
+    )
 
-    # mutating function gene in active region returns False
-    assert genome._mutate_hidden_region(list(dna), 6, active_regions, rng) is False
-    # mutating inactive genes in active region returns True
-    assert genome._mutate_hidden_region(list(dna), 7, active_regions, rng) is True
-    assert genome._mutate_hidden_region(list(dna), 8, active_regions, rng) is True
+    n = 10000
+    n_mutations = []
+    for _ in range(n):
+        dna_old = genome.dna
+        genome.mutate(mutation_rate, rng)
+        n_mutations.append(count_mutations(dna_old, genome.dna))
 
-    # mutating any gene in active region without silent genes returns False
-    assert genome._mutate_hidden_region(list(dna), 9, active_regions, rng) is False
-    assert genome._mutate_hidden_region(list(dna), 10, active_regions, rng) is False
-    assert genome._mutate_hidden_region(list(dna), 11, active_regions, rng) is False
+    assert np.mean(n_mutations) == pytest.approx(n_mutations_mean_expected, rel=0.02)
+    assert np.std(n_mutations) == pytest.approx(n_mutations_std_expected, rel=0.02)
 
 
-def test_mutate_output_region(rng_seed):
+def test_only_silent_mutations(genome_params, mutation_rate, rng_seed):
+    genome = cgp.Genome(**genome_params)
     rng = np.random.RandomState(rng_seed)
-    genome = cgp.Genome(1, 1, 2, 1, None, (cgp.Add,))
-    dna = [
-        ID_INPUT_NODE,
-        ID_NON_CODING_GENE,
-        ID_NON_CODING_GENE,
-        0,
-        0,
-        0,
-        0,
-        0,
-        0,
-        ID_OUTPUT_NODE,
+    genome.randomize(rng)
+
+    only_silent_mutations_0 = genome.mutate(mutation_rate=0, rng=rng)
+    assert only_silent_mutations_0
+
+    only_silent_mutations_1 = genome.mutate(mutation_rate=1, rng=rng)
+    assert not only_silent_mutations_1
+
+    graph = CartesianGraph(genome)
+    active_regions = graph.determine_active_regions()
+
+    gene_to_be_mutated_active = active_regions[-1] * (
+        genome.primitives.max_arity + 1
+    )  # function gene of the 1st active hidden gene, should always be mutable
+    gene_to_be_mutated_non_active = 4 * (
+        genome.primitives.max_arity + 1
+    )  # 4 is a non active gene in this seed
```

Comment:
i would avoid making this depend on the seed, rather hardcode the dna above and avoid `randomize`

Reply:
addressed in [https://github.com/Happy-Algorithms-League/hal-cgp/pull/180/commits/a59116918515c098c719caff22e53237abb7e7f7](url)